test/tests/functional/pbs_multi_sched.py

```diff
@@ -0,0 +1,600 @@
+        a = {'sched_priv': '/var/spool/pbs/sc1_new_priv'}
+        self.server.expect(SCHED, a, id='sc1', max_attempts=10)
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id="sc1")
+        a = {'comment': ''}
+        self.server.expect(SCHED, a, id='sc1', max_attempts=10)
+        self.du.rm(path=os.path.join(pbs_home, 'sc1_new_priv'), recursive=True)
+
+    def test_set_sched_log(self):
+        """
+        Test sched_log can be only set to valid paths
+        and check for appropriate comments
+        """
+        sched = self.server.schedulers['sc1']
+        if not os.path.exists('/var/sched_log_do_not_exist'):
+            self.server.manager(MGR_CMD_SET, SCHED,
+                                {'sched_log': '/var/sched_log_do_not_exist'},
+                                id="sc1")
+        a = {'sched_log': '/var/sched_log_do_not_exist',
+             'comment': 'Unable to change the sched_logs directory',
+             'scheduling': 'False'}
+        self.server.expect(SCHED, a, id='sc1', max_attempts=10)
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id="sc1")
+        pbs_home = self.server.pbs_conf['PBS_HOME']
+        self.du.run_copy(self.server.hostname,
+                         os.path.join(pbs_home, 'sched_logs'),
+                         os.path.join(pbs_home, 'sc1_new_logs'),
+                         recursive=True)
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'sched_log': '/var/spool/pbs/sc1_new_logs'},
+                            id="sc1")
+        a = {'sched_log': '/var/spool/pbs/sc1_new_logs'}
+        self.server.expect(SCHED, a, id='sc1', max_attempts=10)
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id="sc1")
+        a = {'comment': ''}
+        self.server.expect(SCHED, a, id='sc1', max_attempts=10)
+        self.du.rm(path=os.path.join(pbs_home, 'sc1_new_logs'), recursive=True)
+
+    def test_start_scheduler(self):
+        """
+        Test that scheduler wont start without creation from qmgr and
+        appropriate folders created and partition assigned and test
+        states of scheduler in each stage
+        """
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'partition': (DECR, 'P3')}, id="sc3")
+        # Unset sc3's partition to assign to this scheduler
+        pbs_home = self.server.pbs_conf['PBS_HOME']
+        self.server.manager(MGR_CMD_CREATE, SCHED,
+                            id="sc5")
+        a = {'sched_host': self.server.hostname,
+             'sched_port': '15055',
+             'scheduling': 'True'}
+        self.server.manager(MGR_CMD_SET, SCHED, a, id="sc5")
+        sched = self.server.schedulers['sc5']
+        # Try starting without sched_priv and sched_logs
+        ret = self.server.schedulers['sc5'].start(sched_port='15055')
```

Comment:
I was unaware that PTL already had a self.schedulers.  From what I can see self.schedulers is used when -p servers is used.  It needs to be modified anyway because right now it will only be a list of default schedulers.  There won't be any way to get at the non-default schedulers on the other servers.  I suggest that you make self.schedulers a list of dictionaries.  The list will be each of the -p servers, and the dictionary will be a dictionary of schedulers.  So self.schedulers[0]['default'] is self.scheduler.  It's only used in two places in our tests, so it won't be hard to update.

Reply:
I agree with Bhroam, you could just make it a list of dictionaries and it will resolve the problem.

Reply:
Hi @bhroam , @arungrover , So the dictionary of self.schedulers will contain the hostname provided in -p and list of schedulers per host .
But do you think we should call all scheduler functions using self.schedulers[0]['sched_name'] ?
Or is it fine to call self.server.schedulers['sched_name'].func_name(), We are currently using self.server.schedulers dictionary values such log_path and priv_path inside Class Sched functions .

Reply:
I dislike self.server.schedulers because a scheduler is a daemon and not part of the server.  It isn't like a queue which is a server concept.  It is true that the scheduler information is held at the server, but so is mom information held by the vnodes.  We should be using self.schedulers.